2021-Firmware-Bootcamp/Core/Src/main.c

```diff
@@ -0,0 +1,214 @@
+  * the "License"; You may not use this file except in compliance with the
+  * License. You may obtain a copy of the License at:
+  *                        opensource.org/licenses/BSD-3-Clause
+  *
+  ******************************************************************************
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+#include "spi.h"
+#include "tim.h"
+#include "usart.h"
+#include "gpio.h"
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+uint16_t const MAXADC = 1024;
+uint16_t const MAXPWM = 0xFFFF;
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
+  /* USER CODE BEGIN 2 */
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+
+	  uint8_t const TRANS[3] = {0x01,0x40,0x01};
+	  uint8_t recv[3] = {0};
+	  uint16_t const SIZE = 3;
+	  uint32_t const TIMEOUT = 10000;
+
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_RESET);
+	  HAL_SPI_TransmitReceive(&hspi1, TRANS , recv, SIZE, TIMEOUT);
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+
+	  uint16_t adc = (uint16_t) recv[1];
+
+	  adc = adc<<14; // push recv[1] B9 and B8 bits to 15th and 16th bit removing all other unnecessary bits
+	  adc = adc>>7; //push B9 and B8 to 8th and 9th position
+	  adc = adc + (uint16_t) recv[2]; // add the bits from recv 2 as they are only 8 bits they should stitch with recv last 2 bits
```

Comment:
So I think this would work, but it seems unconventional. An equivalent expression that is more conventional is `adc = ((recv[1] << 8) + recv[2]) & 0x3FF` 

So instead of shifting back and forth to get rid of the garbage bits by making them go out of range, you can just use the whole value then use the 0x3FF bit mask to cut them off by ANDing the values together.